### PR TITLE
feat #1092: Validate binary values + precompile them as well

### LINKED ISSUE #1092: Also validate the binary matcher if template validation was requested
Example:
```yaml
matchers:
  - type: binary
     binary:
       - "504B0304"
```

```golang
func (m *Matcher) MatchBinary(corpus string) (bool, []string) {
...
// DecodeString expects that src contains only hexadecimal characters and that src has even length.
// If the input is malformed, DecodeString returns the bytes decoded before the error.
hex, err := hex.DecodeString(binary)
...
}
```

Search for other usages of `DecodeString` and make sure the error is handled correctly.

### PR DESCRIPTION
**Pull request by Ice3man543**
## Proposed changes

Closes #1092 

## Checklist

- [ ] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [ ] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: v2/pkg/operators/matchers/compile.go [hunk 1]
```diff
@@ -43,6 +43,15 @@ func (m *Matcher) CompileMatchers() error {
 		m.regexCompiled = append(m.regexCompiled, compiled)
 	}
 
+	// Compile and validate binary Values in matcher
+	for _, value := range m.Binary {
+		if decoded, err := hex.DecodeString(value); err != nil {
+			return fmt.Errorf("could not hex decode binary: %s", value)
+		} else {
+			m.binaryCompiled = append(m.binaryCompiled, string(decoded))
```
Comment (forgedhallpass) [inline v2/pkg/operators/matchers/compile.go:51]: there is no compliation going on here, hence the field should be renamed

**Comment by forgedhallpass** [comment]
The associated ticket was about validating hex values from templates when the `-validate` flag is used. Instead of doing so, this PR seems to change the functionality itself. Have you checked if content highlighting with the `-debug` flag still works after this change?

**Review by forgedhallpass — CHANGES_REQUESTED** [review]
The associated ticket was about validating hex values from templates when the -validate flag is used. Instead of doing so, this PR seems to change the functionality itself. Have you checked if content highlighting with the -debug flag still works after this change?

File: v2/pkg/operators/matchers/compile.go (hunk 1 above)
Comment (Ice3man543) [reply v2/pkg/operators/matchers/compile.go]: Done

**Comment by Ice3man543** [comment]
Basically what i did was move the binary hex decoding to the `Compile` function which returns errors if it finds an invalid sequence, similar to what we do for regexes. It should not really have an impact on coloring as that part is not touched.

**Comment by forgedhallpass** [comment]
You are right, when I looked at the code I relied on the name of the variable `hexa`, which in fact holds binary data. Also the `Binary` field in the matcher is confusing, because it actually contains the data in hexadecimal format. Another reason why having meaningful variable names is important...

**Review by ehsandeep — APPROVED** [review]
```console
echo 1.1.1.1 | ./nuclei -t a.yaml -validate

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   2.5.4-dev

		projectdiscovery.io

[WRN] Use with caution. You are responsible for your actions.
[WRN] Developers assume no liability and are not responsible for any misuse or damage.
[ERR] Error occurred parsing template test.yaml: could not compile request: could not compile operators: could not compile matcher: could not hex decode binary: 00700000200000002000000
[FTL] Could not run nuclei: an error occurred during templates validation
```

**Review by Mzack9999 — APPROVED** [review]

**Pull request merged**